### PR TITLE
Quml Bulk upload API as part of JIRA key-> SB-25750

### PR DESCRIPTION
**Pull request by shalinichandru12**
<table>
<td>
  <a href="https://project-sunbird.atlassian.net/browse/SB-25750" title="SB-25750" target="_blank">StorySB-25750</a>  Bulk Question Upload API
  </td></table>
  <br />
 

We propose to implement a Bulk Upload API on the backend to enable uploading multiple (to the order of thousands) questions at once onto Diksha.

### Type of change

Please choose appropriate options.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [yes ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### Checklist:

- [yes ] My code follows the style guidelines of this project
- [yes] I have performed a self-review of my own code
- [yes ] I have commented my code, particularly in hard-to-understand areas
- [yes] I have made corresponding changes to the documentation
- [yes] My changes generate no new warnings
- [yes] I have added tests that prove my fix is effective or that my feature works
- [yes] New and existing unit tests pass locally with my changes
- [yes] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/service/kafkaQumlConsumerService.js [hunk 1]
```diff
@@ -0,0 +1,201 @@
+var kafka = require("kafka-node");
+const fetch = require("node-fetch");
+const logger = require('sb_logger_util_v2');
+const loggerService = require('./loggerService');
+const responseCode = messageUtils.RESPONSE_CODE;
```
Comment (venkateshwarans) [inline src/service/kafkaQumlConsumerService.js:6]: we are not importing `messageUtils` anywhere. Please check.

File: src/service/qumlBulkService.js [hunk 2]
```diff
@@ -0,0 +1,153 @@
+const fs = require("fs");
+const fetch = require("node-fetch");
+const { v4: uuidv4 } = require("uuid");
+const KafkaService = require("../helpers/kafkaUtil");
+const logger = require('sb_logger_util_v2');
+const loggerService = require('./loggerService');
+const responseCode = messageUtils.RESPONSE_CODE;
```
Comment (venkateshwarans) [inline src/service/qumlBulkService.js:7]: we are not importing `messageUtils` in this file anywhere. please check.

File: src/service/qumlBulkService.js [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+const bulkUpload = async (req, res) => {
+    const logObject = {
+        traceId : req.headers['x-request-id'] || '',
+        message : programMessages.CONTENT_PUBLISH.INFO
+      }
+  let errorArray = [];
+  let pId = uuidv4();
+  let successArray = [];
+  let qumlData;
+  const fileType = req.files.File.mimetype;
```
Comment (venkateshwarans) [inline src/service/qumlBulkService.js:22]: shouldn't it be `req.body.files.File.mimeType` ? our request will be always under the `body` right?

File: src/routes/qumlBulkRoutes.js [hunk 4]
```diff
@@ -0,0 +1,11 @@
+const qumlBulkUpload = require("../service/qumlBulkService");
+const BASE_URL = '/api/question/v1/';
+
+module.exports = function (app) {
+    app.route(BASE_URL + '/bulkUpload')
```
Comment (venkateshwarans) [inline src/routes/qumlBulkRoutes.js:5]: `BASE_URL` is ending with the `/` and the route endpoint is starting with a `/` . we will ending up accessing the api like `question/v1//bulkUpload`. please fix.

File: src/routes/qumlBulkRoutes.js [hunk 5]
```diff
@@ -0,0 +1,11 @@
+const qumlBulkUpload = require("../service/qumlBulkService");
+const BASE_URL = '/api/question/v1/';
```
Comment (venkateshwarans) [inline src/routes/qumlBulkRoutes.js:2]: Please remove `/api`

File: src/service/kafkaQumlConsumerService.js [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+  let consumer = new Consumer(client, [{ topic: envVariables.SUNBIRD_QUESTION_BULKUPLOAD_TOPIC }], {
+    encoding: "utf8",
+    fromOffset: false,
+  });
+//after kafka consumer receives data from the same topic of the kafka producer.
+  consumer
+    .on("message", function (message) {
+      console.log("------------------RESPONE MESSAGE  ----------------------");
+      const qumlArr = JSON.parse(message.value);
+      loggerService.entryLog(message.value);
```
Comment (venkateshwarans) [inline src/service/kafkaQumlConsumerService.js:24]: `entryLog` will expect two parameters. please fix.

File: src/service/qumlBulkService.js (hunk 3 above)
Comment (shalinichandru12) [reply src/service/qumlBulkService.js:22]: No Since it is in form data it will come as files

File: src/service/qumlBulkService.js [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+
+//question search API function;
+const qumlSearch = (req, res) => {
+  const searchData = req.body;
+  const logObject = {
+    traceId: req.headers["x-request-id"] || "",
+    message: programMessages.QUML_BULKSTATUS.INFO,
+  };
+  loggerService.entryLog("Api to check the status of bulk upload question", logObject);
+  fetch(`${envVariables.baseURL}/action/composite/v3/search`, {
```
Comment (pallakartheekreddy) [inline src/service/qumlBulkService.js:189]: use `api/composite/v1/search`

**Review by pallakartheekreddy — APPROVED** [review]

**Pull request merged**